lm_survey/samplers/async_openai_sampler.py

```diff
@@ -0,0 +1,100 @@
+import sys
+
+import openai
+from openai.error import RateLimitError
+
+import torch
+from aiolimiter import AsyncLimiter
+
+from lm_survey.samplers.base_sampler import BaseSampler, MaybeAwaitable
+
+# Constants for throttling
+OPENAI_RPM = 3000
+
+
+class AsyncOpenAiSampler(BaseSampler):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        if "async-gpt3" in self.model_name:
+            # engine is all text after 'async-gpt3-'
+            self.engine = "-".join(self.model_name.split("-")[2:])
+        else:
+            self.engine = self.model_name
+
+        self._async_limiter = AsyncLimiter(OPENAI_RPM)
+
+        print(f"Using async {self.engine} engine.")
+
+        if openai.api_key is None:
+            raise ValueError("OpenAI API key must be set")
+
+    async def rank_completions(self, prompt, completions):
+        # 100 is the maximum number of log probs we can get.
+        top_log_probs, response = await self.send_prompt(prompt, n_probs=100)
+
+        log_probs = torch.tensor(
+            [
+                top_log_probs.get(completion, -torch.inf)
+                for completion in completions
+            ]
+        )
+
+        normalized_log_probs = torch.nn.functional.log_softmax(log_probs, dim=0)
+
+        completion_log_probs = {
+            completion: normalized_log_prob.item()
+            for completion, normalized_log_prob in zip(
+                completions, normalized_log_probs
+            )
+        }
+
+        return completion_log_probs, response
+
+    async def _throttled_completion(self, **kwargs):
+        while True:
+            # We do this inside of the loop so that retries respect the rate limit too.
+            async with self._async_limiter:
+                try:
+                    return await openai.Completion.acreate(
+                        engine=self.engine, **kwargs
+                    )
+                except RateLimitError:
+                    # TODO: This is not a good way to do logging; we should actually use
+                    # the logging module or something similar.
+                    if self.logger:
+                        self.logger.exception("Rate limited, retrying...")
```

Comment:
`logger.exception` is only used to add context to caught exceptions—almost always in an `except` block. This isn't even really an error. `logger.warning` would probably be more appropriate here.

https://docs.python.org/3/library/logging.html#logging.Logger.exception